cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerLicenseService.java

```diff
@@ -0,0 +1,64 @@
+package com.sequenceiq.cloudbreak.cm;
+
+import java.util.Optional;
+
+import javax.inject.Inject;
+import javax.ws.rs.core.Response;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import com.cloudera.api.swagger.client.ApiClient;
+import com.cloudera.api.swagger.client.ApiException;
+import com.sequenceiq.cloudbreak.auth.altus.GrpcUmsClient;
+import com.sequenceiq.cloudbreak.cm.client.ClouderaManagerClientFactory;
+import com.sequenceiq.cloudbreak.workspace.model.User;
+
+@Service
+public class ClouderaManagerLicenseService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClouderaManagerLicenseService.class);
+
+    @Inject
+    private ClouderaManagerClientFactory clouderaManagerClientFactory;
+
+    @Inject
+    private GrpcUmsClient umsClient;
+
+    public void beginTrialIfNeeded(User user, ApiClient apiClient) throws ApiException {
+        // Begin the Cloudera Manager trial only if UMS is not enabled. Otherwise, we'll be using a
+        // license from UMS.
+        if (needTrialLicense(user.getUserCrn())) {
+            beginTrial(apiClient);
+        } else {
+            LOGGER.info("UMS detected and license key available, skipping trial license.");
+        }
+    }
+
+    private void beginTrial(ApiClient apiClient) throws ApiException {
+        LOGGER.info("Enabling trial license.");
+
+        try {
+            clouderaManagerClientFactory.getClouderaManagerResourceApi(apiClient).beginTrial();
+        } catch (ApiException e) {
+            if (trialAlreadyStarted(e)) {
```

Comment:
Is this catch/continue block basically the real core difference between the previous behavior and the current behavior? Is the implication here is that this is getting run multiple times?

Reply:
The core difference is that trial is started (if needed) before adding MGMT service.  The catch/continue is there to allow both LDAP setup and MGMT service setup to do it independently (since I haven't find a way to query the trial state).  It may also help in case the task is retried -- don't know if that's a real scenario or not.

Reply:
It is, because the flow steps are restartable.